hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/HazelcastSqlValidator.java

```diff
@@ -146,6 +147,28 @@ public void validateCall(SqlCall call, SqlValidatorScope scope) {
         super.validateCall(call, scope);
     }
 
+    @Override
+    public void validateQuery(SqlNode node, SqlValidatorScope scope, RelDataType targetRowType) {
+        super.validateQuery(node, scope, targetRowType);
+
+        if (node instanceof SqlSelect) {
+            // Derive the types for offset-fetch expressions, Calcite doesn't do
+            // that automatically.
+
+            SqlSelect select = (SqlSelect) node;
+
+            SqlNode offset = select.getOffset();
```

Comment:
We can ensure that only supported types are produced here (e.g. integer types), and remove the runtime check in the `Fetch.java`.

Reply:
Addressed in https://github.com/hazelcast/hazelcast/pull/18091/commits/55930f7a3009234321511f5a0b05026122c86400